### PR TITLE
WIP: Dashdb connector

### PR DESCRIPTION
**Pull request by ironsideshu**
The code is very dependent on the DataWorks proxy endpoint. Endpoint dev is blocked currently:  waiting on DataWorks team to implement new feature where given a regex pattern and a S3 bucket/folder, transfer all matching files into a single DashDB table. All matching files need to have the same column structure. ETA for function, according to Arron from Bluemix/DataWorks, is February 28th.

### CODE REVIEW

File: 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+        if shredded_good_run_filepath.nil?
+          if snowplow_tracking_enabled
+            Monitoring::Snowplow.instance.track_load_failed('Cannot find atomic-events directory in shredded/good')
+          end
+
+          raise DatabaseLoadError, 'Cannot find atomic-events directory in shredded/good'
+        end
+
+        altered_enriched_subdirectory = RUN_PATTERN.match(altered_enriched_filepath.key)[1]
```
Comment (alexanderdean) [inline 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb]: Do you need the `altered_enriched_subdirectory = ` there - could just return the right-hand-side of the assignment?

File: 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+        res.value
+      end
+      module_function :create_run_dataworks_transfer
+
+      # Retrieve run folder from shredded/good. Max one folder in there at any given time.
+      #
+      # Parameters:
+      # +config+:: the configuration options
+      Contract Hash => String
```
Comment (alexanderdean) [inline 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb]: Contract missing `snowplow_tracking_enabled`

File: 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+      # Call dataworks_proxy api endpoint with a request body that passes DashDB and S3 details.
+      # The endpoint will create the S3 => DashDB activity and also kick it off.
+      # ** Still waiting on DataWorks team to add new functionality where file pattern matching is allowed.
+      #
+      # Parameters:
+      # +config+:: the configuration options
+      # +target+:: the configuration for this specific target
+      # +run_folder+:: the retrieved run folder from shredded/good
+      Contract Hash, Hash => Num
```
Comment (alexanderdean) [inline 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb]: Contract missing `run_folder`

File: 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+    # Used to find the altered enriched events
+    RUN_PATTERN = /(run=[0-9\-]+\/atomic-events)/
+
+    # Loads the Snowplow atomic event files into DashDB table
+    #
+    # Parameters:
+    # +config+:: the configuration options
+    # +target+:: the configuration for this specific target
+    Contract Hash, Hash => nil
```
Comment (alexanderdean) [inline 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb]: Contract missing `snowplow_tracking_enabled`

**Comment by alexanderdean** [comment]
Cool! Have left some comments. Where is the code that generates the DataWorks load schema based on the events table?

File: 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb (hunk 1 above)
Comment (ironsideshu) [inline 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb]: yup, you're right. copy and pasting has failed me :(

File: 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb (hunk 2 above)
Comment (ironsideshu) [inline 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb]: I'm actually not privy of Contracts so I was following redshift_loader as an example. From that class, in load_events_and_shredded_types, the Contract is Hash, Hash => nil, so I figured that snowplow_tracking_enabled is optional?

File: 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb (hunk 3 above)
Comment (ironsideshu) [inline 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb]: duly noted!

**Comment by ironsideshu** [comment]
That will be done by the custom DataWorks endpoint in the backend. The calls to create and run DataWorks activities _has_ to run through Bluemix right now, hence the reason for the DataWorks proxy. Whenever IBM decides to drop that limitation to DataWorks, I can rework the code.

**Comment by alexanderdean** [comment]
Thanks @ironsideshu . Can you also add the atomic.events table definition for DashDB into this PR? You can create a folder in here: https://github.com/snowplow/snowplow/tree/master/4-storage

File: 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb (hunk 2 above)
Comment (alexanderdean) [inline 4-storage/storage-loader/lib/snowplow-storage-loader/dashdb_loader.rb]: Ah - yep I think that's an error in the Redshift file...

**Comment by alexanderdean** [comment]
As part of this PR, we are going to need documentation that we can add into the wiki on how to setup DashDB, DataWorks etc ready for this load process. Would you be able to add a comment on this PR which starts to list the salient points of this setup process @ironsideshu ?

**Comment by ironsideshu** [comment]
@alexanderdean yup, I can add a tidbit on what is needed to setup DashDB and DataWorks! for the atomic events table definition, a small problem:

Schemas can only be created by DashDB admins, and you don't get such a privilege unless you subscribe to their enterprise tiers (tier plans: https://console.ng.bluemix.net/catalog/services/dashdb). The schema that is assigned to the user is of the same name as the autogenerated username that DashDB assigns upon the creation and binding of the service. And that leads to my question: for the DashDB's atomic-def.sql, how do you want me to set it up--towards the users who are going to use it for the Entry level plan or the Enterprise tiers? Alternatively, I can write a short ruby script that can create the atomic-def.sql given some parameters they input. 

The difference in atomic-def.sql between the Entry and Enterprise tiers are minimal--in Entry, no creation of the atomic schema and the schema that is used is the username (in uppercase), and in Enterprise, the atomic schema would be created and used. For now, I'm just going to assume Enterprise plans.

**Comment by alexanderdean** [comment]
Hi @ironsideshu - sounds good, I would go with the Enterprise level, and add a comment into the file explaining the differences to make (manually) if you are using the Entry level.

**Comment by alexanderdean** [comment]
Hey @ironsideshu - I have just built a version of Scala Hadoop Shred which outputs CSV with all values " enclosed. To try it out:

1. Upgrade to Snowplow R77 - http://snowplowanalytics.com/blog/2016/02/28/snowplow-r77-great-auk-released-with-emr-4.x-series-support/
2. Update your config.yml like so:

```yaml
{% highlight yaml %}
versions:
  hadoop_enrich: 1.6.0
  hadoop_shred: 0.8.0-csv         # WAS 0.8.0
  hadoop_elasticsearch: 0.1.0
{% endhighlight %}
```

Let me know if you have any issues...

**Comment by ironsideshu** [comment]
Closing due to limitations of DataWorks' API. Moved onto the DashDB strategy #2908.

**Pull request closed without merging**